honeybee_energy/properties/aperture.py

```diff
@@ -0,0 +1,122 @@
+# coding=utf-8
+"""Aperture Energy Properties."""
+from ..construction import WindowConstruction
+from ..lib.default.room import generic_costruction_set
+
+
+class ApertureEnergyProperties(object):
+    """Energy Properties for Honeybee Aperture.
+
+    Properties:
+        construction
+        materials
+        r_value
+        u_value
+        r_factor
+        u_factor
+    """
+
+    __slots__ = ('_host', '_construction')
+
+    def __init__(self, host, construction=None):
+        """Initialize Aperture energy properties.
+
+        Args:
+            host: A honeybee_core Aperture object that hosts these properties.
+            construction: An optional Honeybee WindowConstruction object for
+                the aperture. If None, it will be set by the parent Room ConstructionSet
+                or the the Honeybee default generic ConstructionSet.
+        """
+        self._host = host
+        self.construction = construction
+
+    @property
+    def host(self):
+        """Get the Aperture object hosting these properties."""
+        return self._host
+
+    @property
+    def construction(self):
+        """Get or set Aperture Construction.
+
+        If the Construction is not set on the aperture-level, then it will be assigned
+        based on the ConstructionSet assigned to the parent Room.  If there is no
+        parent Room or the the parent Room's ConstructionSet has no construction for
+        the aperture, it will be assigned using the honeybee default generic
+        construction set.
+        """
+        if self._construction:  # set by user
+            return self._construction
+        elif self._host.has_parent and self._host.parent.has_parent:  # set by zone
+            constr_set = self._host.parent.parent.construction_set
+            return constr_set.get_aperture_construction(
+                self._host.boundary_condition.name, self._host.type.name,
+                self._host.parent.type.name)
+        elif self._host.has_parent:  # generic but influenced by parent face
+            return generic_costruction_set.get_aperture_construction(
+                self._host.boundary_condition.name, self._host.type.name,
+                self._host.parent.type.name)
+        else:
+            return generic_costruction_set.get_aperture_construction(
+                self._host.boundary_condition.name, self._host.type.name, 'Wall')
+
+    @construction.setter
+    def construction(self, value):
+        if value is not None:
+            assert isinstance(value, WindowConstruction), \
+                'Expected Window Construction for aperture. Got {}'.format(type(value))
+            value.lock()  # lock editing in case construction has multiple references
+        self._construction = value
+
+    @property
+    def is_construction_set_by_user(self):
+        """Check if construction is set by user."""
+        return self._construction is not None
+
+    @property
+    def materials(self):
```

Comment:
Are you just saying that the "locked construction" error is better for the user to see rather than "materials is read-only"?  I agree with that.  Now that I'm thinking about it more, maybe I should just not expose all of these construction properties on the Aperture/Face level.  We have so many properties now for constructions that there's no way I can expose all of them on the Aperture/Face level.  And it just seems a lot clearer to have them all on the construction object so that people realize that they are all coming from the construction and not somewhere else.

Reply:
I agree with your second suggestion to keep them in construction and one can always get the values from construction if needed.